src/meltano/core/venv_service.py

```diff
@@ -10,93 +10,150 @@
 import shutil
 import subprocess
 import sys
+import typing as t
 from asyncio.subprocess import Process
-from collections import namedtuple
 from collections.abc import Iterable
 from pathlib import Path
 
-from meltano.core.error import AsyncSubprocessError
+from meltano.core.error import AsyncSubprocessError, MeltanoError
 from meltano.core.project import Project
 
+if sys.version_info >= (3, 8):
+    from functools import cached_property
+else:
+    from cached_property import cached_property
+
 logger = logging.getLogger(__name__)
 
-VenvSpecs = namedtuple("VenvSpecs", ("lib_dir", "bin_dir", "site_packages_dir"))
 
-POSIX = VenvSpecs(
-    lib_dir="lib",
-    bin_dir="bin",
-    site_packages_dir=os.path.join(
-        "lib",
-        f"python{'.'.join(str(part) for part in sys.version_info[:2])}",
-        "site-packages",
-    ),
-)
+class VirtualEnv:
+    """Info about a single virtual environment."""
 
-NT = VenvSpecs(
-    lib_dir="Lib",
-    bin_dir="Scripts",
-    site_packages_dir=os.path.join("Lib", "site-packages"),
-)
+    _SUPPORTED_PLATFORMS = {
+        "Linux",
+        "Darwin",
+        "Windows",
+    }
 
-PLATFORM_SPECS = {"Linux": POSIX, "Darwin": POSIX, "Windows": NT}
+    def __init__(
+        self,
+        root: Path,
+        python: Path | str | None = None,
+    ):
+        """Initialize the `VirtualEnv` instance.
 
+        Args:
+            root: The root directory of the virtual environment.
+            python: The path to the Python executable to use, or name to find on the
+                $PATH. Defaults to the Python executable running Meltano.
 
-def venv_platform_specs():
-    """Get virtual environment sub-path info for the current platform.
+        Raises:
+            MeltanoError: The current system is not supported.
+        """
+        self._system = platform.system()
+        if self._system not in self._SUPPORTED_PLATFORMS:
+            raise MeltanoError(f"Platform {self._system!r} not supported.")
+        self.root = root.resolve()
+        self.python_path = self._resolve_python_path(python)
 
-    Raises:
-        Exception: This platform is not supported.
+    @staticmethod
+    def _resolve_python_path(python: Path | str | None) -> str:
+        python_path: str | None = None
 
-    Returns:
-        Virtual environment sub-path info for the current platform.
-    """
-    system = platform.system()
-    try:
-        return PLATFORM_SPECS[system]
-    except KeyError as ex:
-        raise Exception(f"Platform {system!r} not supported.") from ex
+        if python is None:
+            python_path = sys.executable
+        elif isinstance(python, Path):
+            python_path = str(python.resolve())
+        else:
+            python_path = python if os.path.exists(python) else shutil.which(python)
```

Comment:
Would it be possible to fail early and with a better error message when `python` is a number, e.g. `3.8` (a likely user error)? Otherwise users will see:

```
Need help fixing this problem? Visit http://melta.no/ for troubleshooting steps, or to
join our friendly Slack community.

stat: path should be string, bytes, os.PathLike or integer, not ScalarFloat
```

Reply:
Addressed in 2fe14b174a1fa3d5e2632621cb9e090757f4e3ea